mitodl/ocw-studio
# theme assets pipeline

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-studio/issues/938

#### What's this PR do?
In https://github.com/mitodl/ocw-studio/pull/891, we set up a "waterfall" like triggering strategy where publishes would happen in this sequence:

 - Create RC in `ocw-hugo-themes`
 - `ocw-theme-assets`build triggers from merge to `release-candidate`
 - `ocw-www` build triggers from artifacts being deployed to `ol-eng-artifacts`
 - `ocw-www` build creates a static version file upon completion
 - static version file triggers all other course builds

In theory this is a good approach, but in practice we've had a number of issues:

 - The compounding number of static version files in the bucket is making the trigger for it take a longer amount of time with each run
 - Through some analysis, the unzipping of the theme assets tar file is causing a compounding issue with CPU usage

This PR adjusts our whole approach for deploying new theme assets from start to finish:

 - First of all, the pipeline definition for the theme assets build has been moved from `ol-infrastructure` to here in `ocw-studio`.
 - The pipeline is also now templated and deployed using a new management command, `upsert_theme_assets_pipeline`.  It only creates a single pipeline based on the configured `GITHUB_WEBHOOK_BRANCH`, so that everything done on RC uses the `release-candidate` branch, and everything in production uses the `release` branch.
 - The theme assets pipeline has been updated to not zip up the artifacts.  Instead, they are deployed directly to the publish buckets for draft and live, then the `webpack.json` file is uploaded to `ol-eng-artifacts`.
 - The site pipeline definition was updated so that upload of the `webpack.json` file is what triggers the course build, and `ocw-hugo-themes` is pulled in directly from git for the templates
 
#### How should this be manually tested?
 - Set the following in your `.env` file:
```
CONCOURSE_URL=http://concourse:8080
CONCOURSE_PASSWORD=test
CONCOURSE_USERNAME=test
CONCOURSE_TEAM=main
```
 - Start up `ocw-studio` with the local Concourse profile using `docker-compose --profile concourse up`
 - Run `docker-compose run --rm web ./manage.py upsert_theme_asssets_pipeline`
 - Go to http://localhost:8080 in a browser and log into your local Concourse instance with the info above
 - View the theme assets pipeline and make sure it was properly created
 - Create a test course in the `ocw-studio` UI if you haven't already
 - Run `docker-compose run --rm web ./manage.py backpopulate_pipelines --filter <your site id here>`
 - Ensure that your site pipeline was created successfully and that the triggers look correct

If you actually try running any of the pipelines you will likely encounter errors as your local instance doesn't have AWS access to the buckets required.  Further testing will need to be done once this PR is merged to RC.
